x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/ChainingInputStream.java

```diff
@@ -72,6 +76,86 @@
+     * input stream with the bytes from the second input stream. The stream arguments must support
+     * the {@code mark} and {@code reset} operations; otherwise use {@link SequenceInputStream}.
+     *
+     * @param first the input stream supplying the first bytes of the returned {@link ChainingInputStream}
+     * @param second the input stream supplying the bytes after the {@code first} input stream has been exhausted
+     */
+    public static InputStream chain(InputStream first, InputStream second) {
+        if (false == Objects.requireNonNull(first).markSupported()) {
+            throw new IllegalArgumentException("The first component input stream does not support mark");
+        }
+        if (false == Objects.requireNonNull(second).markSupported()) {
+            throw new IllegalArgumentException("The second component input stream does not support mark");
+        }
+        final InputStream firstComponent = new FilterInputStream(first) {
+            @Override
+            public void close() {
+                // silence close
+                // "first" can be reused, and the {@code ChainingInputStream} eagerly closes components after every use
+                // "first" is closed when the returned {@code ChainingInputStream} is closed
+            }
+        };
+        final InputStream secondComponent = new FilterInputStream(second) {
+            @Override
+            public void close() {
+                // silence close
+                // "second" can be reused, and the {@code ChainingInputStream} eagerly closes components after every use
+                // "second" is closed when the returned {@code ChainingInputStream} is closed
+            }
+        };
+        // be sure to remember the start of components because they might be reused
+        firstComponent.mark(Integer.MAX_VALUE);
+        secondComponent.mark(Integer.MAX_VALUE);
+
+        return new ChainingInputStream() {
+
+            @Override
+            InputStream nextComponent(InputStream currentComponentIn) throws IOException {
+                if (currentComponentIn == null) {
+                    // when returning the next component, start from its beginning
+                    firstComponent.reset();
+                    return firstComponent;
+                } else if (currentComponentIn == firstComponent) {
+                    // when returning the next component, start from its beginning
+                    secondComponent.reset();
+                    return secondComponent;
+                } else if (currentComponentIn == secondComponent) {
+                    return null;
+                } else {
+                    throw new IllegalStateException("Unexpected component input stream");
+                }
+            }
+
+            @Override
+            public void close() throws IOException {
+                IOException IOExceptions = null;
+                try {
+                    super.close();
```

Comment:
Why not just use `org.elasticsearch.core.internal.io.IOUtils#close(java.io.Closeable...)` here?

Reply:
Pushed 8e036f0 
Great suggestion, thank you! It slipped my mind.